examples/parallel_generator.py

```diff
@@ -0,0 +1,61 @@
+"""
+========================================
+Returning a generator in joblib.Parallel
+========================================
+
+This example illustrates memory optimization enabled by using
+:class:`joblib.Parallel` to get a generator on the outputs of parallel jobs.
+We first create tasks that return results with large memory footprints.
+We parallelize several of those tasks such that we should observe a high
+memory usage if all of the outputs were to stack in RAM. We show that the
+memory is efficiently managed if we use the generator to perform a reduce
+step that progressively consumes the outputs and keeps the memory at an
+acceptable level. The RAM usage can depend on the behavior of the garbage
+collector, whose behavior can be hard to predict. Here we force it to ensure
+a low memory usage.
+
+"""
+
+##############################################################################
+# Save memory by consuming the outputs of the tasks as fast as possible
+##############################################################################
+#
+# We create a task whose output takes about 15MB of RAM
+
+import time
+import numpy as np
+
+
+def return_big_object(i):
+    time.sleep(1)
+    return i * np.ones((10000, 200), dtype=np.float64)
+
+
+##############################################################################
+# We create a reduce step. The input will be a generator on big objects
+# generated in parallel by several instances of :func:`return_big_object`.
+
+import gc
+
+
+def accumulator_sum(generator):
+    result = 0
+    for value in generator:
+        result += value
+        del value
+        gc.collect()  # make sure to flush value from memory
```

Comment:
This should not be necessary unless the code introduces a reference cycle. Is this the case here?

Reply:
The `del` is not necessary either, if we can remove both, that would make the code cleaner and closer to user code.

If this causes some increase in the memory consumption, this is indeed caused by trailing references that we should track and remove.

Reply:
I did a few run with `mprof run` and both keep the memory usage close to 200 MB. with `return_generator=False` on the other hand on reaches 3 GB by the end of the computation.

Reply:
I updated the example to include a graphic at the end to show the memory usage:
![image](https://user-images.githubusercontent.com/3321081/153062758-1c46b3ee-009d-4c71-ac1c-2c0dccb1f9fa.png)

Reply:
See [the generated example](https://joblib--588.org.readthedocs.build/en/588/auto_examples/parallel_generator.html#returning-a-generator-in-joblib-parallel).

A question though is whether the `MemoryMonitor` class add too much complexity to the example or not...